complete checkout operations

antis81 opened the pull request:
Realization of the basic `CheckoutOperations` concept.

Up to now, libgit2 only provides very low-level support for doing checkouts. Thus, we're mostly on our own here and need to design on how this is realized in our API. 

:recycle: I really like the basic idea of defining `Operations`. To get a good restart, I cleaned up and wiped out some unfinished concepts (i.e. `Reference::checkoutOperation()`).

Some very basic scnenarios work (again) - yay :+1:
- checkout HEAD tree
- checkout any commit from the history
- checkout references (also branches)
- checkout the index

TODO:
- [x] Finish API design.
  - [x] Use new `CheckoutCallbacks` API
  - [x] Implement and use `Operation Providers`
  - [x] Implement a virtual `postCheckout` method
- [x] Update the HEAD pointing to a reference, when needed.
- [ ] Add some API doumentation.
- [x] Some basic tests for repo index are added.
- [x] fixed double free in `StrArray`
- [x] Creation of lightweight and normal tags.
- [x] Added (failing) tests for creation of tags. This need work on its own -> handled in new PR.

:exclamation: :memo: **Hint:** This **_WILL_** break GitWrap API. Thus, some MGV features will need adjustment after merged.

antis81 committed:
CheckoutOperation: renamed flags -> strategy as it better fits the purpose

antis81 committed:
minor code cosmetics

antis81 committed:
fix: libgit2 reposrts progress with total_steps = 0

antis81 committed:
API change: moved branch creation from Commit to a static function in BranchRef

antis81 committed:
CheckoutCallback: initialize checkout notification flags in callback initialization

antis81 committed:
actually initialize with CheckoutCallbacks

antis81 committed:
API change: removed Reference::checkout and Reference::checkoutOperation from the API

These are currently not working and need a general overhaul.

antis81 committed:
API change: execute() returns no more a Result object.

The returned Result is misleading, when using the asynchronous operation API. Instead, the result() method can be used.

antis81 committed:
fixup commit: removed relict from previous checkout callbacks

antis81 committed:
CheckoutOperations: renewed tree and index operations

antis81 committed:
CheckoutOperations: re-implemented BranchOperation as ReferenceOperation

antis81 committed:
updated libgit2

antis81 committed:
implemented ReferenceTreeProvider

antis81 committed:
added internal static helpers to get "validated git pointers"

These functions are likely to be moved in the future, and are more meant as a "good start".

antis81 committed:
made "prepare" and "unprepare" protected virtual

antis81 committed:
minor: removed empty d'tor

antis81 committed:
changed internal architecture to use internal run methods

This has several advantages:
1. Allows "early exit" from within each runCheckout() method (i.e. return on bad result).
2. Implements common things only once
3. Keeps clear and simple code structure by design, without doing anything bad!

TODO:
Implement a virtual "postCheckout" method to do specific things (i.e. "update head to checked out reference", "update index", etc.)

antis81 committed:
put "isRunning" asserts in front and minor cleanup

antis81 committed:
API change: introducing the TreeProvider for TreeCheckoutOperations

antis81 committed:
Finally! Inherited CheckoutReferenceOperation is a CheckoutTreeOperation.

Reached functional parity at this state. Just everything got better!

antis81 commented:
:+1: Everything just got better. With flowers and butterflies :bouquet: :bouquet: - and brand new bugs :beetle: :grin:. Hard work though :sweat_smile:.

@scunz I know you don't have much time, but ... please ... I am excited to hear your thoughts on this one.

scunz commented on libGitWrap/Events/Private/GitEventCallbacks.cpp:
```diff
@@ -213,8 +212,9 @@ namespace Git
+            Q_ASSERT( events );
 
-            qreal progress = 100 / total_steps * completed_steps;
+            qreal progress = total_steps ? (100 / total_steps * completed_steps) : 100;
```
Shouldn't it be 0% progress if no steps are defined?

antis81 replied:
The only sane reason I can think of for reporting with `total_steps=0` is, "when nothing needs to be done". Assuming, this is not a lg2 bug, 100% is correct.

scunz replied:
Think that's okay then.

antis81 replied:
Thx! Got some time for chatting?

scunz replied:
Yepp :)

antis81 replied:
Ok nice, FB then?

(sorry ... I'll delete the last comments)

antis81 replied:
Btw: libgit2 itself stumbled upon this just a few days ago. They set 0 in their example, but I still think 100 is the correct value. I'll simply ask them and maybe change this in a new PR, as I'm about to merge and what's there pretty much looks ok to me. See here: https://github.com/ethomson/libgit2/commit/9af3c4169c51b967579042e392937acbc75da79d

antis81 replied:
Ah forget about it! We don't need that: See eadcaa0d1caa99a7b5576cde52676e28093b05d4

scunz commented:
Just as a heads up: I saw that CMN is trying to remove the `signature` parameters everywhere. The reasoning is sane, but that will be a major API change in lg2 (see their 2808).

antis81 commented:
> Just as a heads up: I saw that CMN is trying to remove the signature parameters everywhere. The reasoning is sane, but that will be a major API change in lg2 (see their 2808).

Thanks for the hint. For GW it's not a big deal though. I'll deprecate our calls to keep compatibility as soon as the PR got merged over there. We use "NULL" for both parameters anyways.

antis81 committed:
fixed invalid pointer on string conversion

antis81 committed:
minor cleanup

antis81 committed:
set flags through unsigned int

antis81 committed:
moved post checkout operations to separate methods

antis81 committed:
set HEAD to checked out reference, when "CheckoutUpdateHEAD" is set

antis81 committed:
fix build bug

antis81 committed:
cleanup

antis81 committed:
adjust to libgit2 api

antis81 committed:
added some tests for repo-index

antis81 committed:
fixed StrArray

antis81 committed:
StrArray can use explicit c'tors

antis81 committed:
replaced macros "GW_D_CHECKED_VOID" and "GW_CD_CHECKED_VOID"

antis81 committed:
cleanup

antis81 committed:
removed over-pedantic checks

An empty paths list is actually valid and the check on a bare repository isdone by libgit2.

antis81 committed:
removed Commit::setAsDetachedHEAD()

Setting the detached HEAD is done via a commit's ObjectId.

antis81 committed:
renamed objectType2gitotype() -> objectType2git()

antis81 committed:
cleanup

antis81 committed:
implemented Commit::ParentProvider and Commit::TreeProvider

antis81 committed:
implemented CheckoutCommitoperation

antis81 committed:
removed Commit::checkout() and Tree::checkout()

antis81 committed:
renamed ObjectId2git_oid() -> ObjectId2git() and some minor cleanup

antis81 committed:
Tag: functions to create and lookup tag objects

Git::Tag are not created, but only looked up from the Git database.

antis81 committed:
Testing: added tests for tag creation (unfinished)

antis81 committed:
Code Quality: added result checks to commit

antis81 committed:
minor cleanup

antis81 committed:
Testing: renamed file Test_commit.cpp -> TestCommit.cpp

antis81 committed:
Code Quality: check result on tag lookup

antis81 committed:
Tags: don't force tag creation by default

antis81 committed:
checkout: introduced doneCheckout() event

antis81 committed:
minor cleanup

antis81 committed:
remote callbacks: adjusted to libgit2 v0.22

This allows us to centralize the mRemoteCallbacks member and in turn improve code quality.

antis81 committed:
remote operations: added result checks

antis81 committed:
checkout progress: report as quint64 type

antis81 committed:
checkout / clone: fixed memory allocation of strings

antis81 committed:
avoid problems with progress reporting and division by zero
